keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,258 @@
+	// to different external versions.
+	Versioned []*CachingVersionedObject
+}
+
+// TODO: Better name is welcome :)
+type CachingVersionedObject struct {
+	// Object in an versioned object.
+	Object Object
+	// Serializations is a list of cached serializations of Object.
+	// Note that a list is needed, because a given versioned object can be
+	// serialized to different media types.
+	//
+	// TODO: Maybe we should be more explicit and just store one for json
+	// and one for protobuf?
+	Serializations []*serializedObject
+}
+
+// TODO: Better name is welcome :)
+type serializedObject struct {
+	once sync.Once
+	// mediaType to which the object is serialized.
+	mediaType string
+	// raw is serialized object.
+	raw []byte
+	// err is error from serialization.
+	err error
+}
+```
+Note that both `CachingObject` and `CachingVersionedObject` should implement
+`runtime.Object` interface.
+
+For those two objects, we will implement `Marshal()`, `MarshallTo()` (for
+protobuf) and `MarshalJSON()` (for json) methods, that:
+- first check if such serialization has already been done, and if so just
+  returning the cached result
+- otherwise performing the serialization, caching the result and returning
+  it to the called
+
+With that, we will need to modify `versioning.codec` so that when encoding
+`CachingObject`, it will perform necessary conversion if needed to the
+`CachingVersionedObject`.
+
+Now the question is what is responsible for creating `CachingObject`.
+We will put that responsibility into `watchCache` - when receiving an
+event via watch from etcd it will be opaquing it into `CachingObject`
+and operating on objects of that type later.
+
+Note that given that objects are already opaqued in `watchCache`, we
+also avoid unnecessary serialization when GET/LIST requests are served
+from cache.
+
+One open questions is how do we decide what exactly we want to cache.
+Caching everything stored in `watchCache` seems expensive from memory
+usage point of view. What we suggest, is allow configuring individual
+`registries` with exact versions that we want to cache and propagate
+that information down to `Cacher`. That would give us ability to switch
+it on for most demanding resources (e.g. only Node, Endpoints and Pod
+ifrom core/v1 group).
+
+Note that based on a [POC][] (slightly different that above design),
+the gains of implementing it include:
+- eliminating kube-apiserver unresponsiveness in case of write of
+a single huge Endpoints object: [#75294#comment-472728088][]
+- ~7% lower cpu-usage
+- XX% less memory allocations
+
+[POC]: https://github.com/kubernetes/kubernetes/pull/60067
+[#75294#comment-472728088]: https://github.com/kubernetes/kubernetes/issues/75294#issuecomment-472728088
+
+### Risks and Mitigations
+
+Given this proposal doesn't introduce any user visible change, the only risk
+are bugs in implementation. Fortunately, the serialization code is widely
+used by all end-to-end tests, so any bugs should be catched by those or
```

Comment:
Sorry, I need to disagree here.  I think that our experience with the watch cache is that subtle bugs can exist for a long time across multiple releases in spite of usage in tests and by end users.

Reply:
I have to agree with David here :/

Reply:
Let's leave that discussion for a bit later...